### PR TITLE
Added support in  the proxy for events as a top level both json and and protobuf, and the forwarding of events onto via protobuf to ingest. Also split the Sink Interface into DSink and ESink so the interfaces can be used individually.

### PR DESCRIPTION
**Pull request by mdubbyap**
[Review on Reviewable](https://reviewable.io/reviews/signalfx/metricproxy/30)

### CODE REVIEW

File: dp/dpsink/sink.go [hunk 1]
```diff
@@ -6,13 +6,25 @@ import (
...
+
+// A ESink is an object that can accept events and do something with them, like forward them
+// to some endpoint
+type ESink interface {
 	AddEvents(ctx context.Context, events []*event.Event) error
 }
 
+// A Sink is an object that can accept datapoints or events and do something with them, like forward them
+// to some endpoint
+type Sink interface {
```
Comment (cep21) [inline dp/dpsink/sink.go]: +1

File: protocol/signalfx/signalfx.go [hunk 2]
```diff
@@ -147,3 +148,43 @@ func NewProtobufDataPointWithType(dp *com_signalfx_metrics_protobuf.DataPoint, m
...
+		pkey := dpdim.GetKey()
+		if pval.StrValue != nil {
+			props[pkey] = pval.GetStrValue()
+		} else if pval.BoolValue != nil {
+			props[pkey] = pval.GetBoolValue()
+		} else if pval.DoubleValue != nil {
+			props[pkey] = pval.GetDoubleValue()
+		} else if pval.IntValue != nil {
+			props[pkey] = pval.GetIntValue()
+		}
```
Comment (cep21) [inline protocol/signalfx/signalfx.go]: else return an error?

File: protocol/signalfx/signalfxforwarder.go [hunk 3]
```diff
@@ -213,6 +198,50 @@ func (connector *Forwarder) coreDatapointToProtobuf(point *datapoint.Datapoint)
...
+	return v
+}
+
+func mapToProperties(properties map[string]interface{}) []*com_signalfx_metrics_protobuf.Property {
+	ret := make([]*com_signalfx_metrics_protobuf.Property, 0, len(properties))
+	for k, v := range properties {
+		if k == "" || v == nil {
+			continue
+		}
+		copyOfK := filterSignalfxKey(string([]byte(k)))
```
Comment (cep21) [inline protocol/signalfx/signalfxforwarder.go]: ??? You convert a string to []byte then the []byte into a string?  Why?

File: protocol/signalfx/signalfxforwarder.go [hunk 4]
```diff
@@ -213,6 +198,50 @@ func (connector *Forwarder) coreDatapointToProtobuf(point *datapoint.Datapoint)
...
+		pv := com_signalfx_metrics_protobuf.PropertyValue{}
+		if ival, ok := v.(int64); ok {
+			pv.IntValue = &ival
+		} else if bval, ok := v.(bool); ok {
+			pv.BoolValue = &bval
+		} else if dval, ok := v.(float64); ok {
+			pv.DoubleValue = &dval
+		} else if sval, ok := v.(string); ok {
+			pv.StrValue = &sval
+		}
```
Comment (cep21) [inline protocol/signalfx/signalfxforwarder.go]: else ?  Error?  or just skip the value?

File: protocol/signalfx/signalfxlistener.go [hunk 5]
```diff
@@ -162,7 +163,7 @@ func datapointProtobufIsInvalidForV1(msg *com_signalfx_metrics_protobuf.DataPoin
 // JSONDecoderV1 creates datapoints out of the v1 JSON definition
 type JSONDecoderV1 struct {
 	TypeGetter MericTypeGetter
-	Sink       dpsink.Sink
+	Sink       dpsink.DSink
```
Comment (cep21) [inline protocol/signalfx/signalfxlistener.go]: Yay!

**Comment by cep21** [comment]
looks good.  Just a few comments.

**Comment by cep21** [comment]
Reviewed 3 of 7 files at r1.
Review status: 3 of 7 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/signalfx/metricproxy/30)*

**Comment by cep21** [comment]
Reviewed 2 of 7 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/signalfx/metricproxy/30)*

**Comment by cep21** [comment]
Still not sure about the no error code return.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/signalfx/metricproxy/30#-:-K5CDph1mbLRQ6zgSB3v:1658311007)*

**Comment by mdubbyap** [comment]
How about now?

File: protocol/signalfx/signalfxforwarder.go (hunk 4 above)
Comment (cep21) [inline protocol/signalfx/signalfxforwarder.go]: What about this place?

Comment (cep21) [inline protocol/signalfx/signalfxforwarder.go]: I think it's reasonable to `continue` in this situation.  You don't want to send in a PropertyValue that has no Value set.

**Comment by cep21** [comment]
Just on comment on line 256

**Pull request merged**